### PR TITLE
[GSoC] Added support of using `StateSpace` with `Parallel` and `MIMOParallel`.

### PR DESCRIPTION
**Pull request by abhiphile**
Added support for `StateSpace` in `Parallel` and `MIMOParallel` connection.
#### Brief description of what is fixed or changed
 - `TransferFunction` & `TransferFunctionMatrix` can be used in parallel with `StateSpace`.
 -  For the connection having `StateSpace` in it, `.doit()` will return the equivalent `StateSpace` object.

#### Release Notes

* physics.control
  * Added support to use `StateSpace` with `Parallel` and `MIMOParallel`.

### CODE REVIEW

**Comment by abhiphile** [comment]
I've tested my test cases by using `python-control`.

## SISO Docs

```
A1 = np.array([[-1]])  
B1 = np.array([[1]])  
C1 = np.array([[-1]])  
D1 = np.array([1])  
A2 = np.array([[0]])  
B2 = np.array([[1]])  
C2 = np.array([[1]])  
D2 = np.array([[0]])
  
ss1 = control.StateSpace(A1, B1, C1, D1)  
ss2 = control.StateSpace(A2, B2, C2, D2)  
p5 = control.parallel(ss1, ss2)
print(p5)

<StateSpace>: sys[29]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']
States (2): ['x[0]', 'x[1]']

A = [[-1.  0.]
     [ 0.  0.]]

B = [[1.]
     [1.]]

C = [[-1.  1.]]

D = [[1.]]
```

## MIMO Docs
```
A1 = np.array([[4, 1], [2, -3]])  
B1 = np.array([[5, 2], [-3, -3]])  
C1 = np.array([[2, -4], [0, 1]])  
D1 = np.array([[3, 2], [1, -1]])  
A2 = np.array([[-3, 4, 2], [-1, -3, 0], [2, 5, 3]])  
B2 = np.array([[1, 4], [-3, -3], [-2, 1]])  
C2 = np.array([[4, 2, -3], [1, 4, 3]])  
D2 = np.array([[-2, 4], [0, 1]])
ss1 = control.ss(A1, B1, C1, D1)
ss2 = control.ss(A2, B2, C2, D2)
num = [[[1, 0], [1]], [[1], [1, 0]]]
den = [[[1, 1], [1, 0]], [[1, 1], [1, 2]]]
tf2 = control.TransferFunction(num, den)

print(tf2)
<TransferFunction>: sys[30]
Inputs (2): ['u[0]', 'u[1]']
Outputs (2): ['y[0]', 'y[1]']


Input 1 to output 1:
  s
-----
s + 1

Input 1 to output 2:
  1
-----
s + 1

Input 2 to output 1:
1
-
s

Input 2 to output 2:
  s
-----
s + 2

par1 = control.parallel(ss1, ss2)

print(par1)
<StateSpace>: sys[33]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']
States (2): ['x[0]', 'x[1]']

A = [[-1.  0.]
     [ 0.  0.]]

B = [[1.]
     [1.]]

C = [[-1.  1.]]

D = [[1.]]
```

## SISO Test Case
```
import numpy as np
import control

a1 = np.array([[0, 1], [1, 0]])
b1 = np.array([[0], [1]])
c1 = np.array([[0, 1]])
d1 = np.array([[0]])
ss1 = control.ss(a1, b1, c1, d1)
a2 = np.array([[1, 0], [0, 1]])
b2 = np.array([[1], [0]])
c2 = np.array([[1, 0]])
d2 = np.array([[1]])
ss2 = control.ss(a2, b2, c2, d2)
par1 = control.parallel(ss1, ss2)
print(par1)

<StateSpace>: sys[2]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']
States (4): ['x[0]', 'x[1]', 'x[2]', 'x[3]']

A = [[0. 1. 0. 0.]
     [1. 0. 0. 0.]
     [0. 0. 1. 0.]
     [0. 0. 0. 1.]]

B = [[0.]
     [1.]
     [1.]
     [0.]]

C = [[0. 1. 1. 0.]]

D = [[1.]]


tf1 = control.TransferFunction(np.array([1, 0]), np.array([1, 1]))

par_tf = control.parallel(ss1, tf1)
print(par_tf)

<StateSpace>: sys[23]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']
States (3): ['x[0]', 'x[1]', 'x[2]']

A = [[ 0.  1.  0.]
     [ 1.  0.  0.]
     [ 0.  0. -1.]]

B = [[0.]
     [1.]
     [1.]]

C = [[ 0.  1. -1.]]

D = [[1.]]
```

## MIMO Tests
```
a3 = np.array([[4, 1], [2, -3]])  
b3 = np.array([[5, 2], [-3, -3]])  
c3 = np.array([[2, -4], [0, 1]])  
d3 = np.array([[3, 2], [1, -1]])  
a4 = np.array([[-3, 4, 2], [-1, -3, 0], [2, 5, 3]])  
b4 = np.array([[1, 4], [-3, -3], [-2, 1]])  
c4 = np.array([[4, 2, -3], [1, 4, 3]])  
d4 = np.array([[-2, 4], [0, 1]])

ss3 = control.StateSpace(a3, b3, c3, d3)
ss4 = control.StateSpace(a4, b4, c4, d4)
par4 = control.parallel(ss3, ss4)
print(par4)

<StateSpace>: sys[26]
Inputs (2): ['u[0]', 'u[1]']
Outputs (2): ['y[0]', 'y[1]']
States (5): ['x[0]', 'x[1]', 'x[2]', 'x[3]', 'x[4]']

A = [[ 4.  1.  0.  0.  0.]
     [ 2. -3.  0.  0.  0.]
     [ 0.  0. -3.  4.  2.]
     [ 0.  0. -1. -3.  0.]
     [ 0.  0.  2.  5.  3.]]

B = [[ 5.  2.]
     [-3. -3.]
     [ 1.  4.]
     [-3. -3.]
     [-2.  1.]]

C = [[ 2. -4.  4.  2. -3.]
     [ 0.  1.  1.  4.  3.]]

D = [[1. 6.]
     [1. 0.]]
```

File: sympy/physics/control/tests/test_lti.py [hunk 1]
```diff
@@ -1885,3 +1885,120 @@ def test_StateSpace_series():
...
+                                         (TransferFunction(1, s + 1, s), TransferFunction(s, s + 2, s)))),
+                 StateSpace(Matrix([
+                     [4, 1],
+                     [2, -3]]), Matrix([
+                     [5, 2],
+                     [-3, -3]]), Matrix([
+                     [2, -4],
+                     [0, 1]]), Matrix([
+                     [3, 2],
+                     [1, -1]])))
```
Comment (faze-geek) [inline sympy/physics/control/tests/test_lti.py:2004]: We should definitely improve the format of writing tests like these. Readability of test files should be maintained.

**Comment by abhiphile** [comment]
@faze-geek @Sc0rpi0n101 Can you review this PR ?

**Comment by faze-geek** [comment]
Outputs look good to me. I believe you are planning to add a couple of tests on symbolic matrices in a new PR (for all methods) ? If yes, I approve of this one. 

I'll requested a review for SY Lee as well.

**Comment by abhiphile** [comment]
> Outputs look good to me. I believe you are planning to add a couple of tests on symbolic matrices in a new PR (for all methods) ? 
Yes I will add some tests on symbolic matrices.

**Review by faze-geek — APPROVED** [review]

**Pull request merged**